README.md
<h1>R package 'pivot' (version 3.0)</h1>
<h2>Formant extraction based on vowel trajectory dynamics</h2>

<p>by <a href="http://www.aaronalbin.com/">Aaron Albin</a> (with <a href="http://www.ling.rochester.edu/wilrankinen/">Wil Rankinen</a>)</p>

<h3>What the package does:</h3>
<p>
	This package implements the method for extracting measurements from the <a href="http://en.wikipedia.org/wiki/Formant">formant</a> tracks of vowels as described in <a href="http://scitation.aip.org/content/asa/journal/jasa/136/4/10.1121/1.4899475">Albin & Rankinen (2014)</a>.
	Every formant point sampled within a vowel is considered as a possible "pivot" (i.e. turning point), with monophthongs modeled as having one pivot and diphthongs modeled as having two pivots.
	The optimal pivot for the vowel is then determined by fitting regression lines to the formant trajectory and comparing the goodness-of-fit of these lines to the raw formant data.
	In addition to formants, the package can also be applied to any one-dimensional time-series of acoustic measurements (F0, intensity, nasality, breathiness, etc.).
</p>
<p>
	For further details about the method, see the poster presented at the 168th meeting of the Acoustical Society of America (Indianapolis, Indiana; 10/27/2014 ~ 10/31/2014) about the method.
	A PDF copy of the poster can be downloaded from <a href="https://www.academia.edu/8975464/A_pivot_model_for_extracting_formant_measurements_based_on_vowel_trajectory_dynamics">Academia.edu</a> and <a href="https://www.researchgate.net/publication/267390239_A_pivot_model_for_extracting_formant_measurements_based_on_vowel_trajectory_dynamics">ResearchGate</a>.
</p>

<h3>Installation:</h3>
1. Click on the green "Clone or download" button at the top of this page, then click "Download ZIP" on the bottom-right of the pop-up window. Save the .zip file somewhere on your computer.
2. Unzip the .zip file to somewhere convenient (e.g. your desktop).
<ul>
	<li>After this step, the .zip file is no longer needed and can be deleted.</li>
</ul>
3. Open R.
<ul>
	<li>If you are running Windows, you must open R as an administrator. To do so, go to the icon for R on your start menu ('start screen' in Windows 8) or desktop, right-click it, and select "Run as administrator". Click "Yes" to the User Account Control window that pops up.</li>
</ul>
4. Find the path to the 'pivot' folder inside the unzipped folder from step 2. (Crucially, you need to find the path to the folder named 'pivot' and <b>not</b> the higher-level folder named 'pivot-master'.) If you unzipped to your desktop in step 2, the path will be something like the following:
<ul>
	<li><u>For Windows</u>: "C:/Users/MyUsername/Desktop/pivot-master/pivot/"</li>
	<li><u>For Mac</u>: "/Users/MyUsername/Desktop/pivot-master/pivot/"</li>
</ul>
5. Run the following line of code, adjusting the path for the <i>pkgs</i> argument as needed so it points to path you determined in Step 4.
<ul>
	<li><i>install.packages(pkgs="(path from Step 4)", repos=NULL, type="source")</i></li>
</ul>

<p>If everything works correctly, you should see something like the following in the R console:</p>

<p><i>
	* installing *source* package 'pivot' ...<br/>
	** R<br/>
	** data<br/>
	** inst<br/>
	** preparing package for lazy loading<br/>
	** help<br/>
	*** installing help indices<br/>
	** building package indices<br/>
	** testing if installed package can be loaded<br/>
	* DONE (pivot)
</i></p>

<h3>Usage</h3>
<p>
You can load this package as you would any other, e.g. by including <i>library("pivot")</i> at the top of your R script.
Once loaded, the pivot analysis of formant data is performed by two functions - (singular) <i>pivot()</i> for the one-pivot/monophthong analysis and (plural) <i>pivots()</i> for the two-pivot/diphthong analysis.
Both have two arguments - 'f1' and 'f2' - corresponding to the F1 and F2 tracks across the duration of a given vowel token.
To analyze other one-dimensional acoustic data (F0, intensity, nasality, breathiness, etc.), simply provide the relevant vector as 'f1' and leave 'f2' unspecified.
To read more about these functions (including several examples illustrating how to use them), type <i>help("pivot")</i> or <i>?pivot</i> at the R console.
To go through all of these examples at once, run the command <i>example("pivot")</i>.
</p>

<h3>Please cite as:</h3>
<p>
	Albin, A., and W. Rankinen (2014). <a href="http://scitation.aip.org/content/asa/journal/jasa/136/4/10.1121/1.4899475">A "pivot" model for extracting formant measurements based on vowel trajectory dynamics</a>. <i>Journal of the Acoustical Society of America</i>, <i>136</i>(4), 2082.
</p>
<p>(Note that this information can be retrieved at any time by typing <i>citation("pivot")</i> at the R console.)</p>

<h3>License:</h3>
<p>This package is released under the <a href="http://www.gnu.org/copyleft/gpl.html">GNU General Public License, Version 3</a>. Once the package has been installed, a copy of this license can be viewed at any time by typing <i>RShowDoc("LICENSE",package="pivot")</i> at the R console.</p>

<h3>Note:</h3>
<p>This package is unrelated to the <a href="http://cran.r-project.org/web/packages/PivotalR/index.html">PivotalR</a> package on CRAN</p>


pivot/NEWS
NEWS in package 'pivot'

3.0 06-04-2016
    o	Added in support for F3 (i.e. three-dimensional analyses)

2.0 01-18-2015
    o	Added in support for one-dimensional analyses (e.g. F0)
    o	(Temporarily) removed non-functional code for weighting regression fits

1.0 10-27-2014
    o	Initial public release


pivot/R/pivots.r
# pivots.r
# Aaron Albin (with Wil Rankinen)

# This function applies a two-pivot model to a diphthong's formant trajectory.
# The formant trajectory can be provided either in terms of F1 and F2 (the function's 'two-dimensional' application) or in terms of F1, F2, and F3 (its 'three-dimensional' application).
# Alternatively, the function can be used to apply a two-pivot model to any one-dimensional string of numbers (e.g. F0, intensity, etc.)

# Dependencies: FindBestPivots(), RunLinearModel()
# Used in: None

################################################################################################
# This program is free software. You can redistribute it and/or modify it under the terms of   #
#    version 3 of the GNU General Public License as published by the Free Software Foundation. #
# This program is distributed in the hope that it will be useful, but without any warranty -   #
#    without even the implied warranty of merchantability or fitness for a particular purpose. #
# For details on the GNU General Public License, see: http://www.gnu.org/licenses/             #
################################################################################################

pivots = function(
f1, # A vector of positive numbers (e.g. the F1 track across the entire duration of a given vowel token)
f2=NULL, # Same as 'F1' argument except for F2. Leave as NULL for a one-dimensional analysis.
f3=NULL  # Same as 'F1' argument except for F3. Leave as NULL for a one-dimensional analysis of if you only want to include F1 and F2 in the analysis (i.e. a two-dimensional analysis).
){ # End argument list; begin function definition

############################
# Determine dimensionality #
############################

# Determine whether the analysis is one-dimensional (e.g. fundamental frequency), two-dimensional (e.g. F1 and F2), or three-dimensional (e.g. F1, F2, and F3).
if( is.null(f1)){stop("F1 cannot be NULL.")}
if( is.null(f2) & !is.null(f3)){stop("If F3 is provided (i.e. non-NULL), F2 must also be provided.")}
if( is.null(f2) &  is.null(f3)){nDimensions=1} # F2 and F3 are both NULL -> 1 dimension
if(!is.null(f2) &  is.null(f3)){nDimensions=2} # F2 is provided but F3 is NULL -> 2 dimensions
if(!is.null(f2) & !is.null(f3)){nDimensions=3} # F2 and F3 are both provided -> 3 dimensions

####################
# Input validation #
####################

# If two or three dimensions, check to make sure all vectors are the same length
if(nDimensions==2){
	if( length(f1) != length(f2) ){ stop("The vectors for 'f1' and 'f2' must be the same length.") }
}
if(nDimensions==3){
	if( ! ( ( length(f1)==length(f2) ) & ( length(f2)==length(f3) ) ) ){ stop("'f1', 'f2', and 'f3' must all be the same length.") }
}

# Once this has been verified, store the length based on the F1 vector (arbitrarily)
nPoints = length(f1)

# Make sure there are sufficient points to do the analysis
# Since the pivot itself is omitted from the regressions, and since 2 points are required for a regression, this means 6 or more points are required (the 2 pivot-points themselves plus 2 additional points on each side).
if( nPoints<6 ){ stop("There must be 6 or more points to apply a two-pivot model.") }

###########################
# Determine optimal pivot #
###########################

PercentsAcross = seq(from=0,to=1,length=nPoints)
BestPivots = FindBestPivots( nPoints=nPoints, PercentsAcross=PercentsAcross, F1F=f1, F2F=f2, F3F=f3)

PivotA_Index = BestPivots[1]
PivotB_Index = BestPivots[2]
PivotA_PercentAcross = PercentsAcross[PivotA_Index]
PivotB_PercentAcross = PercentsAcross[PivotB_Index]
LeftTimeBump = -1*PivotA_PercentAcross # e.g. if PivotA_Index is 0.55, would be -0.55
RightTimeBump = 1-PivotB_PercentAcross # e.g. if PivotB_Index is 0.55, would be  0.45

###############
# F1 analysis #
###############

LeftRegression_F1  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotA_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f1)
RightRegression_F1 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotB_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f1)
LeftSlope_F1  = as.numeric( coefficients(LeftRegression_F1 ) )
RightSlope_F1 = as.numeric( coefficients(RightRegression_F1) )
PivotA_F1 = f1[PivotA_Index]
PivotB_F1 = f1[PivotB_Index]
Onset_F1  = PivotA_F1 + LeftSlope_F1  * LeftTimeBump
Offset_F1 = PivotB_F1 + RightSlope_F1 * RightTimeBump

###############
# F2 analysis #
###############

if(nDimensions==2 | nDimensions==3){
LeftRegression_F2  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotA_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f2)
RightRegression_F2 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotB_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f2)
LeftSlope_F2  = as.numeric( coefficients(LeftRegression_F2 ) )
RightSlope_F2 = as.numeric( coefficients(RightRegression_F2) )
PivotA_F2 = f2[PivotA_Index]
PivotB_F2 = f2[PivotB_Index]
Onset_F2  = PivotA_F2 + LeftSlope_F2  * LeftTimeBump
Offset_F2 = PivotB_F2 + RightSlope_F2 * RightTimeBump
} # End 'if two- or three-dimensional'

###############
# F3 analysis #
###############

if(nDimensions==3){
LeftRegression_F3  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotA_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f3)
RightRegression_F3 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotB_Index, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f3)
LeftSlope_F3  = as.numeric( coefficients(LeftRegression_F3 ) )
RightSlope_F3 = as.numeric( coefficients(RightRegression_F3) )
PivotA_F3 = f3[PivotA_Index]
PivotB_F3 = f3[PivotB_Index]
Onset_F3  = PivotA_F3 + LeftSlope_F3  * LeftTimeBump
Offset_F3 = PivotB_F3 + RightSlope_F3 * RightTimeBump
} # End 'if three-dimensional'

#####################
# Return the result #
#####################

if(nDimensions==1){ # i.e. if one-dimensional, label the first column simply 'x'
Output = data.frame( x       = c( Onset_F1, PivotA_F1,            PivotB_F1,            Offset_F1 ),
                     percent = c( 0,        PivotA_PercentAcross, PivotB_PercentAcross, 1         )  )
}
if(nDimensions==2){ # If two-dimensional, label the first two columns 'f1' and 'f2'
Output = data.frame( f1      = c( Onset_F1, PivotA_F1,            PivotB_F1,            Offset_F1 ),
                     f2      = c( Onset_F2, PivotA_F2,            PivotB_F2,            Offset_F2 ),
                     percent = c( 0,        PivotA_PercentAcross, PivotB_PercentAcross, 1         )  ) # 0=Onset_PercentAcross, 1=Offset_PercentAcross
}
if(nDimensions==3){
Output = data.frame( f1      = c( Onset_F1, PivotA_F1,            PivotB_F1,            Offset_F1 ),
                     f2      = c( Onset_F2, PivotA_F2,            PivotB_F2,            Offset_F2 ),
                     f3      = c( Onset_F3, PivotA_F3,            PivotB_F3,            Offset_F3 ),
                     percent = c( 0,        PivotA_PercentAcross, PivotB_PercentAcross, 1         )  ) # 0=Onset_PercentAcross, 1=Offset_PercentAcross
}

rownames(Output) <- c("onset","pivotA","pivotB","offset")
# Note: In version 1 of this code, 'offglide' was erroneously used instead of 'offset'

return(Output)

} # End definition of function 'pivots()'


pivot/R/FindBestPivot.r
# FindBestPivot.r
# Aaron Albin (with Wil Rankinen)

# This function takes in the information about a formant contour and determines the best pivot for it (in a one-pivot/monophthong model)

# Dependency: RunLinearModel()
# Used in: pivot()

################################################################################################
# This program is free software. You can redistribute it and/or modify it under the terms of   #
#    version 3 of the GNU General Public License as published by the Free Software Foundation. #
# This program is distributed in the hope that it will be useful, but without any warranty -   #
#    without even the implied warranty of merchantability or fitness for a particular purpose. #
# For details on the GNU General Public License, see: http://www.gnu.org/licenses/             #
################################################################################################

FindBestPivot = function(nPoints, PercentsAcross, F1F, F2F, F3F){

# Determine whether the analysis is one-dimensional (e.g. fundamental frequency), two-dimensional (e.g. F1 and F2), or three-dimensional (e.g. F1, F2, and F3).
if( is.null(F2F) ){ # i.e. F2 is empty, hence regardless of F3 it must be a one-dimensional analysis
nDimensions=1
}else{
	if( is.null(F3F) ){ # i.e. F2 is *not* NULL but F3 is, hence it is a two-dimensional analysis
		nDimensions=2
	}else{ # Neither F2 nor F3 are NULL, hence it's a three-dimensional analysis
		nDimensions=3
	} # End 'if/else third dimension is null'
} # End 'if/else second dimension is null'

GoodnessOfFit = rep(NA,times=nPoints)
for( EachPivotCandidate in 2:(nPoints-1) ){ # Note that 2 and nPoints-1 are included because only two points are needed for an lm fit

# Run the linear models for F1
# This is always done (since F1 is included regardless of the 1/2/3 dimension distinction)
F1LinearModel_Left  = RunLinearModel(Type="Left",  PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F1F)
F1LinearModel_Right = RunLinearModel(Type="Right", PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F1F)
# Concatenate the F1 residuals for the left and right sides of model fit
F1Residuals = as.numeric( c( residuals( F1LinearModel_Left  ),
                             residuals( F1LinearModel_Right ) ) )

# If this is a two- or three- dimensional analysis, do the same for F2
if(nDimensions==2 | nDimensions==3 ){
F2LinearModel_Left  = RunLinearModel(Type="Left",  PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F2F)
F2LinearModel_Right = RunLinearModel(Type="Right", PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F2F)
F2Residuals = as.numeric( c( residuals( F2LinearModel_Left  ),
                             residuals( F2LinearModel_Right ) ) )
} # End 'if two- or three-dimensional'

# If this is a three-dimensional analysis, do the same for F3
if(nDimensions==3){
F3LinearModel_Left  = RunLinearModel(Type="Left",  PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F3F)
F3LinearModel_Right = RunLinearModel(Type="Right", PivotCandidateIndex=EachPivotCandidate, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=F3F)
F3Residuals = as.numeric( c( residuals( F3LinearModel_Left  ),
                             residuals( F3LinearModel_Right ) ) )
} # End 'if three-dimensional'

# Determine the goodness-of-fit
if(nDimensions==1){ # If one-dimensional, use just the the median absolute residual for F1
	GoodnessOfFit[EachPivotCandidate] <- median(abs(F1Residuals))
}
if(nDimensions==2){ # If two-dimensional, use the median absolute residual for F1 plus that of F2
	GoodnessOfFit[EachPivotCandidate] <- median(abs(F1Residuals)) + median(abs(F2Residuals))
}
if(nDimensions==3){ # If three-dimensional, use the median absolute residual for F1 plus that of F2 plus that of F3
	GoodnessOfFit[EachPivotCandidate] <- median(abs(F1Residuals)) + median(abs(F2Residuals)) + median(abs(F3Residuals))
}

} # End 'each pivot candidate' loop

# Since bigger residuals are a bad thing, find the model with the smallest goodness-of-fit value
BestFit = min( GoodnessOfFit, na.rm=TRUE )
Indices_BestFit = which( GoodnessOfFit == BestFit )
BestPivotCandidate = Indices_BestFit[length(Indices_BestFit)]
# If there is an exact tie (which should be very rare), arbitrarily take the last candidate with that goodness-of-fit value.

# Output the index corresponding to the best pivot
return(BestPivotCandidate)

} # End definition of function 'FindBestPivot()'


pivot/R/FindBestPivots.r
# FindBestPivots.r
# Aaron Albin (with Wil Rankinen)

# This function takes in the information about a formant contour and determines the best pair of pivots for it (in a two-pivot/diphthong model)
#    The core calculation is the median absolute residual (from the beginning to pivot candidate X) plus (from the end to pivot candidate Y).
#    Everything else is just linear interpolation between pivot candidates X and Y.

# Dependency: InterpolationResiduals()
# Used in: pivots()

################################################################################################
# This program is free software. You can redistribute it and/or modify it under the terms of   #
#    version 3 of the GNU General Public License as published by the Free Software Foundation. #
# This program is distributed in the hope that it will be useful, but without any warranty -   #
#    without even the implied warranty of merchantability or fitness for a particular purpose. #
# For details on the GNU General Public License, see: http://www.gnu.org/licenses/             #
################################################################################################

FindBestPivots = function(nPoints, PercentsAcross, F1F, F2F, F3F){

# Determine whether the analysis is one-dimensional (e.g. fundamental frequency), two-dimensional (e.g. F1 and F2), or three-dimensional (e.g. F1, F2, and F3).
if( is.null(F2F) ){ # i.e. F2 is empty, hence regardless of F3 it must be a one-dimensional analysis
nDimensions=1
}else{
	if( is.null(F3F) ){ # i.e. F2 is *not* NULL but F3 is, hence it is a two-dimensional analysis
		nDimensions=2
	}else{ # Neither F2 nor F3 are NULL, hence it's a three-dimensional analysis
		nDimensions=3
	} # End 'if/else third dimension is null'
} # End 'if/else second dimension is null'

# ----- F1 -----
# This is always done (since F1 is included regardless of the 1/2/3 dimension distinction)

F1LeftRegressionResiduals_Unappended = lapply( X=2:nPoints, FUN=function(EachPoint){
   Regression = lm( I(F1F[1:EachPoint]-F1F[EachPoint])~I(PercentsAcross[1:EachPoint]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F1LeftRegressionResiduals = append( list(NULL), F1LeftRegressionResiduals_Unappended ) # Add empty first componend

F1RightRegressionResiduals_Unappended = lapply( X=1:(nPoints-1), FUN=function(EachPoint){
   Regression = lm(I(F1F[EachPoint:nPoints]-F1F[EachPoint])~I(PercentsAcross[EachPoint:nPoints]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F1RightRegressionResiduals = append(F1RightRegressionResiduals_Unappended, list(NULL) ) # Add empty last component

# ----- F2 -----
# If this is a two- or three- dimensional analysis, do the same for F2

if(nDimensions==2 | nDimensions==3 ){

F2LeftRegressionResiduals_Unappended = lapply( X=2:nPoints, FUN=function(EachPoint){
   Regression = lm( I(F2F[1:EachPoint]-F2F[EachPoint])~I(PercentsAcross[1:EachPoint]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F2LeftRegressionResiduals = append( list(NULL), F2LeftRegressionResiduals_Unappended ) # Add empty first component

F2RightRegressionResiduals_Unappended = lapply( X=1:(nPoints-1), FUN=function(EachPoint){
   Regression = lm(I(F2F[EachPoint:nPoints]-F2F[EachPoint])~I(PercentsAcross[EachPoint:nPoints]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F2RightRegressionResiduals = append(F2RightRegressionResiduals_Unappended, list(NULL) ) # Add empty last component

} # End 'if two- or three-dimensional'


# ----- F3 -----
# If this is a three- dimensional analysis, do the same for F3

if( nDimensions==3 ){

F3LeftRegressionResiduals_Unappended = lapply( X=2:nPoints, FUN=function(EachPoint){
   Regression = lm( I(F3F[1:EachPoint]-F3F[EachPoint])~I(PercentsAcross[1:EachPoint]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F3LeftRegressionResiduals = append( list(NULL), F3LeftRegressionResiduals_Unappended ) # Add empty first component

F3RightRegressionResiduals_Unappended = lapply( X=1:(nPoints-1), FUN=function(EachPoint){
   Regression = lm(I(F3F[EachPoint:nPoints]-F3F[EachPoint])~I(PercentsAcross[EachPoint:nPoints]-PercentsAcross[EachPoint])+0 )
   return( as.numeric(residuals(Regression)) )
   } # End definition of function
) # End call to 'lapply()'
F3RightRegressionResiduals = append(F3RightRegressionResiduals_Unappended, list(NULL) ) # Add empty last component

} # End 'if three-dimensional'

# -----

# Create a matrix of linear interpolation residuals
AllCombinations = t( combn(2:(nPoints-1), 2) ) # Crucially, 1 and nPoints are excluded
LeftHalves = AllCombinations[,1]
RightHalves = AllCombinations[,2]

# Draw on the above two results vectors to re-assemble (and duplicate) the same information once for each combination
F1ResidualsFromLeft = lapply(LeftHalves,FUN=function(x){return(F1LeftRegressionResiduals[[x]])})
F1ResidualsFromRight = lapply(RightHalves,FUN=function(x){return(F1RightRegressionResiduals[[x]])})
F1InterpolationResiduals <- apply(AllCombinations, MARGIN=1, FUN=InterpolationResiduals, FormantVector=F1F)
# Note that 'FormantVector' is an argument to 'InterpolationResiduals()'
CollectedResiduals_F1 = rbind(F1ResidualsFromLeft,F1InterpolationResiduals,F1ResidualsFromRight)
FlattenedResiduals_F1 = apply(CollectedResiduals_F1,MARGIN=2,FUN=function(x){as.numeric(unlist(x))})

# If two- or three-dimensional, do the same for F2
if( nDimensions==2 | nDimensions==3 ){
F2ResidualsFromLeft = lapply(LeftHalves,FUN=function(x){return(F2LeftRegressionResiduals[[x]])})
F2ResidualsFromRight = lapply(RightHalves,FUN=function(x){return(F2RightRegressionResiduals[[x]])})
F2InterpolationResiduals <- apply(AllCombinations, MARGIN=1, FUN=InterpolationResiduals, FormantVector=F2F)
CollectedResiduals_F2 = rbind(F2ResidualsFromLeft,F2InterpolationResiduals,F2ResidualsFromRight)
FlattenedResiduals_F2 = apply(CollectedResiduals_F2,MARGIN=2,FUN=function(x){as.numeric(unlist(x))})
} # End 'if two-dimensional'

# If three-dimensional, do the same for F3
if( nDimensions==3 ){
F3ResidualsFromLeft = lapply(LeftHalves,FUN=function(x){return(F3LeftRegressionResiduals[[x]])})
F3ResidualsFromRight = lapply(RightHalves,FUN=function(x){return(F3RightRegressionResiduals[[x]])})
F3InterpolationResiduals <- apply(AllCombinations, MARGIN=1, FUN=InterpolationResiduals, FormantVector=F3F)
CollectedResiduals_F3 = rbind(F3ResidualsFromLeft,F3InterpolationResiduals,F3ResidualsFromRight)
FlattenedResiduals_F3 = apply(CollectedResiduals_F3,MARGIN=2,FUN=function(x){as.numeric(unlist(x))})
} # End 'if three-dimensional'

# Summarize the results in an evaluation metric (namely, the median absolute residual)
if(nDimensions==1){
	EvaluationMetric = apply(X=FlattenedResiduals_F1,MARGIN=2,FUN=function(x){median(abs(x))})
}
if(nDimensions==2){
	EvaluationMetric =
		apply(X=FlattenedResiduals_F1,MARGIN=2,FUN=function(x){median(abs(x))}) +
		apply(X=FlattenedResiduals_F2,MARGIN=2,FUN=function(x){median(abs(x))})
	# Needs to be apply() and not lapply() because FlattenedResiduals (F1 and F2) is a matrix, not a vector or list. 
}
if(nDimensions==3){
	EvaluationMetric =
		apply(X=FlattenedResiduals_F1,MARGIN=2,FUN=function(x){median(abs(x))}) +
		apply(X=FlattenedResiduals_F2,MARGIN=2,FUN=function(x){median(abs(x))}) +
		apply(X=FlattenedResiduals_F3,MARGIN=2,FUN=function(x){median(abs(x))})
}

# Output two indices corresponding to the best two pivots
BestPairIndex = which.min(EvaluationMetric)
BestPivotPairing = AllCombinations[BestPairIndex,]
return(BestPivotPairing)

} # End definition of function 'FindBestPivots()'


pivot/R/pivot.r
# pivot.r
# Aaron Albin (with Wil Rankinen)

# This function applies a one-pivot model to a monophthong's formant trajectory.
# The formant trajectory can be provided either in terms of F1 and F2 (the function's 'two-dimensional' application) or in terms of F1, F2, and F3 (its 'three-dimensional' application).
# Alternatively, the function can be used to apply a one-pivot model to any one-dimensional string of numbers (e.g. F0, intensity, etc.)

# Dependencies: FindBestPivot(), RunLinearModel()
# Used in: None

################################################################################################
# This program is free software. You can redistribute it and/or modify it under the terms of   #
#    version 3 of the GNU General Public License as published by the Free Software Foundation. #
# This program is distributed in the hope that it will be useful, but without any warranty -   #
#    without even the implied warranty of merchantability or fitness for a particular purpose. #
# For details on the GNU General Public License, see: http://www.gnu.org/licenses/             #
################################################################################################

pivot = function(
f1, # A vector of positive numbers (e.g. the F1 track across the entire duration of a given vowel token)
f2=NULL, # Same as 'F1' argument except for F2. Leave as NULL for a one-dimensional analysis.
f3=NULL  # Same as 'F1' argument except for F3. Leave as NULL for a one-dimensional analysis of if you only want to include F1 and F2 in the analysis (i.e. a two-dimensional analysis).
){ # End argument list; begin function definition

############################
# Determine dimensionality #
############################

# Determine whether the analysis is one-dimensional (e.g. fundamental frequency), two-dimensional (e.g. F1 and F2), or three-dimensional (e.g. F1, F2, and F3).
if( is.null(f1)){stop("F1 cannot be NULL.")}
if( is.null(f2) & !is.null(f3)){stop("If F3 is provided (i.e. non-NULL), F2 must also be provided.")}
if( is.null(f2) &  is.null(f3)){nDimensions=1} # F2 and F3 are both NULL -> 1 dimension
if(!is.null(f2) &  is.null(f3)){nDimensions=2} # F2 is provided but F3 is NULL -> 2 dimensions
if(!is.null(f2) & !is.null(f3)){nDimensions=3} # F2 and F3 are both provided -> 3 dimensions

####################
# Input validation #
####################

# If two or three dimensions, check to make sure all vectors are the same length
if(nDimensions==2){
	if( length(f1) != length(f2) ){ stop("The vectors for 'f1' and 'f2' must be the same length.") }
}
if(nDimensions==3){
	if( ! ( ( length(f1)==length(f2) ) & ( length(f2)==length(f3) ) ) ){ stop("'f1', 'f2', and 'f3' must all be the same length.") }
}

# Once this has been verified, store the length based on the F1 vector (arbitrarily)
nPoints = length(f1)

# Make sure there are sufficient points to do the analysis
# Since the pivot itself is omitted from the regressions, and since 2 points are required for a regression, this means 5 or more points are required.
if( nPoints<5 ){ stop("There must be 5 or more points to apply a one-pivot model.") }

###########################
# Determine optimal pivot #
###########################

PercentsAcross = seq(from=0,to=1,length=nPoints)
PivotIndex = FindBestPivot(nPoints=nPoints, PercentsAcross=PercentsAcross, F1F=f1, F2F=f2, F3F=f3)
Pivot_PercentAcross = PercentsAcross[PivotIndex]

###############
# F1 analysis #
###############

LeftRegression_F1  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f1)
RightRegression_F1 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f1)
LeftSlope_F1  = as.numeric( coefficients(LeftRegression_F1 ) )
RightSlope_F1 = as.numeric( coefficients(RightRegression_F1) )
Pivot_F1 = f1[PivotIndex]
LeftTimeBump_F1 = -1*Pivot_PercentAcross # e.g. if pivot is 0.55, would be -0.55
Onset_F1 = Pivot_F1 + LeftSlope_F1 * LeftTimeBump_F1
RightTimeBump_F1 = 1-Pivot_PercentAcross # e.g. if pivot is 0.55, would be 0.45
Offset_F1 = Pivot_F1 + RightSlope_F1 * RightTimeBump_F1

###############
# F2 analysis #
###############

if( nDimensions==2 | nDimensions==3 ){
LeftRegression_F2  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f2)
RightRegression_F2 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f2)
LeftSlope_F2  = as.numeric( coefficients(LeftRegression_F2 ) )
RightSlope_F2 = as.numeric( coefficients(RightRegression_F2) )
Pivot_F2 = f2[PivotIndex]
LeftTimeBump_F2 = -1*Pivot_PercentAcross # e.g. if pivot is 0.55, we want this to be -0.55
Onset_F2 = Pivot_F2 + LeftSlope_F2 * LeftTimeBump_F2
RightTimeBump_F2 = 1-Pivot_PercentAcross # e.g. if pivot is 0.55, we want this to be 0.45
Offset_F2 = Pivot_F2 + RightSlope_F2 * RightTimeBump_F2
} # End 'if two- or three-dimensional'

###############
# F3 analysis #
###############

if( nDimensions==3 ){
LeftRegression_F3  = RunLinearModel(Type="Left",  PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f3)
RightRegression_F3 = RunLinearModel(Type="Right", PivotCandidateIndex=PivotIndex, PercentsAcross=PercentsAcross, nPoints=nPoints, FormantVector=f3)
LeftSlope_F3  = as.numeric( coefficients(LeftRegression_F3 ) )
RightSlope_F3 = as.numeric( coefficients(RightRegression_F3) )
Pivot_F3 = f3[PivotIndex]
LeftTimeBump_F3 = -1*Pivot_PercentAcross # e.g. if pivot is 0.55, we want this to be -0.55
Onset_F3 = Pivot_F3 + LeftSlope_F3 * LeftTimeBump_F3
RightTimeBump_F3 = 1-Pivot_PercentAcross # e.g. if pivot is 0.55, we want this to be 0.45
Offset_F3 = Pivot_F3 + RightSlope_F3 * RightTimeBump_F3
} # End 'if three-dimensional'

#####################
# Return the result #
#####################

if(nDimensions==1){ # if one-dimensional, label the first column simply 'x'
Output = data.frame( x       = c( Onset_F1,  Pivot_F1,            Offset_F1),
                     percent = c( 0,         Pivot_PercentAcross, 1        ) )	
}
if(nDimensions==2){ # If two-dimensional, label the first two columns 'f1' and 'f2'
Output = data.frame( f1      = c( Onset_F1,  Pivot_F1,            Offset_F1),
                     f2      = c( Onset_F2,  Pivot_F2,            Offset_F2),
                     percent = c( 0,         Pivot_PercentAcross, 1        ) ) # 0=Onset_PercentAcross, 1=Offset_PercentAcross
}
if(nDimensions==3){ # If three-dimensional, label the first three columns 'f1', 'f2', and 'f3'
Output = data.frame( f1      = c( Onset_F1,  Pivot_F1,            Offset_F1),
                     f2      = c( Onset_F2,  Pivot_F2,            Offset_F2),
                     f3      = c( Onset_F3,  Pivot_F3,            Offset_F3),
                     percent = c( 0,         Pivot_PercentAcross, 1        ) ) # 0=Onset_PercentAcross, 1=Offset_PercentAcross
}

rownames(Output) <- c("onset","pivot","offset")

return(Output)

} # End definition of function 'pivot()'
